tests/python/test_matrix.py

```diff
@@ -32,66 +41,109 @@ def test_python_scope_matrix_operations():
 # or Python-scope tensor accesses.
 # ideally we should use pytest.fixture to parameterize the tests
 # over explicit loops
+@pytest.mark.parametrize('ops', vector_operation_types)
 @ti.host_arch_only
-def test_python_scope_vector_tensor_add():
+def test_python_scope_vector_tensor(ops):
     t1 = ti.Vector(2, dt=ti.i32, shape=())
     t2 = ti.Vector(2, dt=ti.i32, shape=())
-    a, b = test_vector_arrays
-    t1[None], t2[None] = a, b
-
-    # TODO: hook Matrix.Proxy to redirect to at + Matrix.__add__
-    c = t1.at(None) + t2.at(None)
-    assert np.allclose(c.to_numpy(), a + b)
-
-
-@ti.host_arch_only
-def test_python_scope_vector_tensor_sub():
-    t1 = ti.Vector(2, dt=ti.i32, shape=())
-    t2 = ti.Vector(2, dt=ti.i32, shape=())
-    a, b = test_vector_arrays
-    t1[None], t2[None] = a, b
+    a, b = test_vector_arrays[:2]
+    t1[None], t2[None] = a.tolist(), b.tolist()
 
-    # TODO: hook Matrix.Proxy to redirect to at + Matrix.__sub__
-    c = t1.at(None) - t2.at(None)
-    assert np.allclose(c.to_numpy(), a - b)
+    c = ops(t1[None].value, t2[None].value)
+    assert np.allclose(c.to_numpy(), ops(a, b))
 
 
+@pytest.mark.parametrize('ops', vector_operation_types)
 @ti.host_arch_only
-def test_python_scope_matrix_tensor_add():
+def test_python_scope_matrix_tensor(ops):
     t1 = ti.Matrix(2, 2, dt=ti.i32, shape=())
     t2 = ti.Matrix(2, 2, dt=ti.i32, shape=())
-    a, b = test_matrix_arrays
+    a, b = test_matrix_arrays[:2]
     # ndarray not supported here
     t1[None], t2[None] = a.tolist(), b.tolist()
 
-    # TODO: hook Matrix.Proxy to redirect to at + Matrix.__add__
-    c = t1.at(None) + t2.at(None)
+    c = ops(t1[None].value, t2[None].value)
     print(c)
 
-    assert np.allclose(c.to_numpy(), a + b)
+    assert np.allclose(c.to_numpy(), ops(a, b))
 
 
 @ti.host_arch_only
-def test_python_scope_matrix_tensor_sub():
-    t1 = ti.Matrix(2, 2, dt=ti.i32, shape=())
-    t2 = ti.Matrix(2, 2, dt=ti.i32, shape=())
-    a, b = test_matrix_arrays
-    # ndarray not supported here
-    t1[None], t2[None] = a.tolist(), b.tolist()
+def test_constant_matrices():
+    print(ti.cos(ti.math.pi / 3))
+    print(-ti.Vector([2, 3]))
+    print(ti.cos(ti.Vector([2, 3])))
+    print(ti.max(2, 3))
+    print(ti.max(2, ti.Vector([3, 4, 5])))
+    print(ti.Vector([2, 3]) + ti.Vector([3, 4]))
+    print(ti.atan2(ti.Vector([2, 3]), ti.Vector([3, 4])))
+    print(ti.Matrix([[2, 3], [4, 5]]) @ ti.Vector([2, 3]))
+    v = ti.Vector([3, 4])
+    w = ti.Vector([5, -12])
+    r = ti.Vector([1, 2, 3, 4])
+    s = ti.Matrix([[1, 2], [3, 4]])
+    print(v.normalized())
+    print(v.cross(w))
+    w.y = v.x * w[0]
+    r.x = r.y
+    r.y = r.z
+    r.z = r.w
+    r.w = r.x
+    print(w)
+    s[0, 1] = 2
+    print(s[0, 1])
```

Comment:
Apart from should we also `assert` their results?

If the code is intentionally designed this way just to test if `print` works, please leave a comment.

Reply:
Yes, I added these just to make sure Python-scope functions don't crash Taichi, and we're good enough if so. A good `assert`'ed test is really hard to be systematically (like I did in `test_element_wise`) can be added iapr?

Reply:
I see. I understand `print` is hard to test systematically for now. Given the title is `test_constant_matrices`, should we test if the return values are correct using `asserts`?

Reply:
Added now, although not covered all functions, but can be a good walk around before we come to a systematic test.